In [210]:
import pandas as pd
import numpy as np

In [211]:
from sklearn.impute import SimpleImputer
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
from sklearn.impute import KNNImputer
from sklearn.preprocessing import LabelEncoder

In [212]:
#para añadir continente
import pycountry_convert as pc

In [213]:
# para poder visualizar todas las columnas de los DataFrames
pd.set_option('display.max_columns', None) 

In [214]:
#Abro y creo un dataFrame con la información de los vuelos
df_Calidad_vida = pd.read_csv('./files/Quality_of_Life.csv')

In [215]:
df_Calidad_vida.head()

country  Purchasing Power Value Purchasing Power Category  \
0    Afghanistan                   32.15                'Very Low'   
1  Aland Islands                  125.01               'Very High'   
2        Albania                   42.82                     'Low'   
3       Alderney                    0.00                       NaN   
4        Algeria                   27.60                'Very Low'   

   Safety Value Safety Category  Health Care Value Health Care Category  \
0         25.33           'Low'              24.24                'Low'   
1         71.81          'High'              79.72               'High'   
2         55.52      'Moderate'              48.21           'Moderate'   
3         83.79     'Very High'             100.00          'Very High'   
4         47.54      'Moderate'              54.43           'Moderate'   

   Climate Value Climate Category  Cost of Living Value  \
0           0.00              NaN                 21.08   
1           0.00              NaN                 53.44   
2          86.43      'Very High'                 40.85   
3           0.00              NaN                  0.00   
4          94.82      'Very High'                 25.31   

  Cost of Living Category Property Price to Income Value  \
0              'Very Low'                            7.8   
1                   'Low'                           5.33   
2                   'Low'                          14.88   
3                     NaN                            0.0   
4              'Very Low'                           21.7   

  Property Price to Income Category  Traffic Commute Time Value  \
0                             'Low'                       56.17   
1                             'Low'                       19.05   
2                            'High'                       36.74   
3                               NaN                        5.00   
4                       'Very High'                       45.09   

  Traffic Commute Time Category  Pollution Value Pollution Category  \
0                   'Very High'            84.44        'Very High'   
1                    'Very Low'            18.05         'Very Low'   
2                    'Moderate'            77.25             'High'   
3                    'Very Low'             1.72         'Very Low'   
4                        'High'            63.87             'High'   

  Quality of Life Value Quality of Life Category  
0                   0.0                      NaN  
1                   0.0                      NaN  
2            ': 104.16'                    'Low'  
3                   0.0                      NaN  
4             ': 98.83'               'Very Low'

In [216]:
df_spain = df_Calidad_vida[df_Calidad_vida["country"] == "Spain"]
df_spain

country  Purchasing Power Value Purchasing Power Category  Safety Value  \
199   Spain                  101.41                    'High'         63.51   

    Safety Category  Health Care Value Health Care Category  Climate Value  \
199          'High'              77.26               'High'          94.04   

    Climate Category  Cost of Living Value Cost of Living Category  \
199      'Very High'                 44.41                   'Low'   

    Property Price to Income Value Property Price to Income Category  \
199                           8.28                        'Moderate'   

     Traffic Commute Time Value Traffic Commute Time Category  \
199                       26.76                    'Very Low'   

     Pollution Value Pollution Category Quality of Life Value  \
199             35.7              'Low'            ': 184.67'   

    Quality of Life Category  
199              'Very High'

In [217]:
df_Calidad_vida.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 236 entries, 0 to 235
Data columns (total 19 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   country                            236 non-null    object 
 1   Purchasing Power Value             236 non-null    float64
 2   Purchasing Power Category          190 non-null    object 
 3   Safety Value                       236 non-null    float64
 4   Safety Category                    234 non-null    object 
 5   Health Care Value                  236 non-null    float64
 6   Health Care Category               221 non-null    object 
 7   Climate Value                      236 non-null    float64
 8   Climate Category                   114 non-null    object 
 9   Cost of Living Value               236 non-null    float64
 10  Cost of Living Category            191 non-null    object 
 11  Property Price to Income Value     236 non-null    object 

In [218]:
df_Calidad_vida.isnull().sum()


country                                0
Purchasing Power Value                 0
Purchasing Power Category             46
Safety Value                           0
Safety Category                        2
Health Care Value                      0
Health Care Category                  15
Climate Value                          0
Climate Category                     122
Cost of Living Value                   0
Cost of Living Category               45
Property Price to Income Value         0
Property Price to Income Category     21
Traffic Commute Time Value             0
Traffic Commute Time Category         29
Pollution Value                        0
Pollution Category                    10
Quality of Life Value                  0
Quality of Life Category             122
dtype: int64

In [219]:
nulos = df_Calidad_vida.isnull().sum()
porcentaje_nulos = (nulos / df_Calidad_vida.shape[0]) * 100
porcentaje_nulos

country                               0.000000
Purchasing Power Value                0.000000
Purchasing Power Category            19.491525
Safety Value                          0.000000
Safety Category                       0.847458
Health Care Value                     0.000000
Health Care Category                  6.355932
Climate Value                         0.000000
Climate Category                     51.694915
Cost of Living Value                  0.000000
Cost of Living Category              19.067797
Property Price to Income Value        0.000000
Property Price to Income Category     8.898305
Traffic Commute Time Value            0.000000
Traffic Commute Time Category        12.288136
Pollution Value                       0.000000
Pollution Category                    4.237288
Quality of Life Value                 0.000000
Quality of Life Category             51.694915
dtype: float64

In [220]:
df_Calidad_vida.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 236 entries, 0 to 235
Data columns (total 19 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   country                            236 non-null    object 
 1   Purchasing Power Value             236 non-null    float64
 2   Purchasing Power Category          190 non-null    object 
 3   Safety Value                       236 non-null    float64
 4   Safety Category                    234 non-null    object 
 5   Health Care Value                  236 non-null    float64
 6   Health Care Category               221 non-null    object 
 7   Climate Value                      236 non-null    float64
 8   Climate Category                   114 non-null    object 
 9   Cost of Living Value               236 non-null    float64
 10  Cost of Living Category            191 non-null    object 
 11  Property Price to Income Value     236 non-null    object 

Purchasing Power Category  categorico - moda
Safety Category categorico - moda
Health Care Category  categorico - moda
Climate Category categorico - moda
Cost of Living Category  categorico - moda
Property Price to Income Category categorico - moda
Traffic Commute Time Category categorico - moda
Pollution Category categorico - moda
Quality of Life Category categorico - moda

In [221]:
columnas_object_con_nulos = df_Calidad_vida[df_Calidad_vida.columns[df_Calidad_vida.isnull().any()]].select_dtypes(include = "O").columns

In [222]:
columnas_object_con_nulos

Index(['Purchasing Power Category', 'Safety Category', 'Health Care Category',
       'Climate Category', 'Cost of Living Category',
       'Property Price to Income Category', 'Traffic Commute Time Category',
       'Pollution Category', 'Quality of Life Category'],
      dtype='object')

In [223]:
for col in columnas_object_con_nulos:
    print(f"La distribución de las categorías (incluyendo nulos) para la columna", col)
    display(df_Calidad_vida[col].value_counts(dropna=False, normalize=True) * 100)  # Convertimos a porcentaje
    print("........................")
for col in columnas_object_con_nulos:
    print(f"La distribución de las categorías (incluyendo nulos) para la columna", col)
    display(df_Calidad_vida[col].value_counts(dropna=False, normalize=True) * 100)  # Convertimos a porcentaje
    print("........................")

La distribución de las categorías (incluyendo nulos) para la columna Purchasing Power Category


Purchasing Power Category
'Very Low'     26.694915
'Low'          19.491525
NaN            19.491525
'Very High'    16.525424
'Moderate'      8.898305
'High'          8.898305
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Safety Category


Safety Category
'Moderate'     42.372881
'High'         30.932203
'Low'          17.372881
'Very High'     6.355932
'Very Low'      2.118644
NaN             0.847458
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Health Care Category


Health Care Category
'High'         39.406780
'Moderate'     38.559322
'Low'          11.016949
NaN             6.355932
'Very High'     4.237288
'Very Low'      0.423729
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Climate Category


Climate Category
NaN            51.694915
'Very High'    24.576271
'High'         17.372881
'Moderate'      5.932203
'Low'           0.423729
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Cost of Living Category


Cost of Living Category
'Very Low'     37.288136
'Low'          27.966102
NaN            19.067797
'Moderate'      8.050847
'High'          4.237288
'Very High'     3.389831
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Property Price to Income Category


Property Price to Income Category
'Very High'    31.355932
'High'         17.796610
'Moderate'     17.796610
'Low'          12.711864
'Very Low'     11.440678
NaN             8.898305
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Traffic Commute Time Category


Traffic Commute Time Category
'Very Low'     26.271186
'Low'          24.576271
'Moderate'     18.644068
'High'         13.559322
NaN            12.288136
'Very High'     4.661017
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Pollution Category


Pollution Category
'High'         35.593220
'Moderate'     20.762712
'Low'          16.949153
'Very High'    13.983051
'Very Low'      8.474576
NaN             4.237288
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Quality of Life Category


Quality of Life Category
NaN            51.694915
'Very High'    13.983051
'Very Low'     11.016949
'Moderate'     10.593220
'Low'           8.050847
'High'          4.661017
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Purchasing Power Category


Purchasing Power Category
'Very Low'     26.694915
'Low'          19.491525
NaN            19.491525
'Very High'    16.525424
'Moderate'      8.898305
'High'          8.898305
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Safety Category


Safety Category
'Moderate'     42.372881
'High'         30.932203
'Low'          17.372881
'Very High'     6.355932
'Very Low'      2.118644
NaN             0.847458
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Health Care Category


Health Care Category
'High'         39.406780
'Moderate'     38.559322
'Low'          11.016949
NaN             6.355932
'Very High'     4.237288
'Very Low'      0.423729
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Climate Category


Climate Category
NaN            51.694915
'Very High'    24.576271
'High'         17.372881
'Moderate'      5.932203
'Low'           0.423729
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Cost of Living Category


Cost of Living Category
'Very Low'     37.288136
'Low'          27.966102
NaN            19.067797
'Moderate'      8.050847
'High'          4.237288
'Very High'     3.389831
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Property Price to Income Category


Property Price to Income Category
'Very High'    31.355932
'High'         17.796610
'Moderate'     17.796610
'Low'          12.711864
'Very Low'     11.440678
NaN             8.898305
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Traffic Commute Time Category


Traffic Commute Time Category
'Very Low'     26.271186
'Low'          24.576271
'Moderate'     18.644068
'High'         13.559322
NaN            12.288136
'Very High'     4.661017
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Pollution Category


Pollution Category
'High'         35.593220
'Moderate'     20.762712
'Low'          16.949153
'Very High'    13.983051
'Very Low'      8.474576
NaN             4.237288
Name: proportion, dtype: float64

........................
La distribución de las categorías (incluyendo nulos) para la columna Quality of Life Category


Quality of Life Category
NaN            51.694915
'Very High'    13.983051
'Very Low'     11.016949
'Moderate'     10.593220
'Low'           8.050847
'High'          4.661017
Name: proportion, dtype: float64

........................


In [224]:
df_Calidad_vida["Climate Category"].unique()

array([nan, "'Very High'", "'High'", "'Moderate'", "'Low'"], dtype=object)

In [225]:
df_Calidad_vida["Climate Value"].unique()

array([ 0.  , 86.43, 94.82, 98.28, 63.42, 93.8 , 77.4 , 91.4 , 64.31,
       71.29, 64.37, 86.03, 80.48, 88.87, 94.31, 82.76, 59.43, 55.41,
       90.21, 79.61, 96.69, 99.48, 89.23, 90.87, 79.1 , 82.47, 67.82,
       94.51, 91.98, 86.6 , 64.28, 96.82, 71.89, 58.87, 89.71, 88.53,
       83.3 , 67.03, 91.79, 99.89, 83.64, 78.74, 68.81, 64.1 , 63.76,
       70.99, 46.98, 87.83, 93.78, 90.78, 71.58, 85.26, 89.05, 39.78,
       99.79, 76.84, 20.22, 58.11, 74.7 , 94.74, 96.54, 67.76, 82.62,
       81.7 , 56.55, 97.44, 87.51, 76.91, -3.54, 89.02, 86.09, 92.64,
       87.42, 95.46, 60.75, 76.3 , 68.5 , 67.22, 71.93, 67.84, 92.24,
       97.69, 61.23, 75.9 , 97.31, 36.03, 77.  , 43.6 , 99.74, 45.98,
       83.23, 57.45, 80.72, 76.11, 95.25, 68.39, 94.04, 59.11, 76.49,
       79.24, 84.38, 71.28, 69.76, 67.55, 93.24, 93.26, 98.41, 70.69,
       45.84, 87.41, 79.14, 98.04, 69.79, 71.24, 96.76])

In [226]:
df_Calidad_vida["Climate Value"].value_counts()

Climate Value
0.00     122
77.00      1
97.31      1
75.90      1
61.23      1
        ... 
89.71      1
58.87      1
71.89      1
96.82      1
96.76      1
Name: count, Length: 115, dtype: int64

columnas a rellenar con "Unknown":
Purchasing Power Category            19.491525
Climate Category                     51.694915
Cost of Living Category              19.067797
Traffic Commute Time Category        12.288136
Quality of Life Category             51.694915

Columnas a rellenar con "moda"
Safety Category                       0.847458
Health Care Category                  6.355932
Property Price to Income Category     8.898305
Pollution Category                    4.237288

In [227]:
columnas_modificar_NAN = ["Purchasing Power Category", "Climate Category", "Cost of Living Category", "Traffic Commute Time Category", "Quality of Life Category"]

df_Calidad_vida[columnas_modificar_NAN] = df_Calidad_vida[columnas_modificar_NAN].fillna("Unknown")


In [228]:
df_Calidad_vida.isnull().sum()

country                               0
Purchasing Power Value                0
Purchasing Power Category             0
Safety Value                          0
Safety Category                       2
Health Care Value                     0
Health Care Category                 15
Climate Value                         0
Climate Category                      0
Cost of Living Value                  0
Cost of Living Category               0
Property Price to Income Value        0
Property Price to Income Category    21
Traffic Commute Time Value            0
Traffic Commute Time Category         0
Pollution Value                       0
Pollution Category                   10
Quality of Life Value                 0
Quality of Life Category              0
dtype: int64

In [229]:
columnas_modificar_moda = ["Safety Category","Health Care Category","Property Price to Income Category","Pollution Category"]
for col in columnas_modificar_moda:
    moda = df_Calidad_vida[col].mode()[0]  # Obtener la moda (primer valor en caso de empate)
    df_Calidad_vida[col] = df_Calidad_vida[col].fillna(moda)

In [230]:
df_Calidad_vida.isnull().sum()

country                              0
Purchasing Power Value               0
Purchasing Power Category            0
Safety Value                         0
Safety Category                      0
Health Care Value                    0
Health Care Category                 0
Climate Value                        0
Climate Category                     0
Cost of Living Value                 0
Cost of Living Category              0
Property Price to Income Value       0
Property Price to Income Category    0
Traffic Commute Time Value           0
Traffic Commute Time Category        0
Pollution Value                      0
Pollution Category                   0
Quality of Life Value                0
Quality of Life Category             0
dtype: int64

In [231]:
df_Calidad_vida.describe()

Purchasing Power Value  Safety Value  Health Care Value  Climate Value  \
count              236.000000    236.000000         236.000000     236.000000   
mean                55.573305     55.274449          54.731568      37.598178   
std                 52.008245     16.914298          20.607381      40.851542   
min                  0.000000      0.000000           0.000000      -3.540000   
25%                 16.340000     43.857500          45.807500       0.000000   
50%                 42.930000     54.635000          57.150000       0.000000   
75%                 85.940000     68.132500          68.447500      79.332500   
max                281.830000    100.000000         100.000000      99.890000   

       Cost of Living Value  Traffic Commute Time Value  Pollution Value  
count            236.000000                  236.000000       236.000000  
mean              37.526314                   28.492966        54.266186  
std               26.026565                   17.347242        25.853695  
min                0.000000                    0.000000         0.000000  
25%               24.550000                   17.100000        35.700000  
50%               36.895000                   29.845000        59.765000  
75%               51.090000                   38.870000        73.740000  
max              137.370000                  100.000000       106.900000

In [232]:
df_Calidad_vida["Climate Value"]= df_Calidad_vida["Climate Value"].abs()

In [233]:
df_Calidad_vida.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
231    False
232    False
233    False
234    False
235    False
Length: 236, dtype: bool

In [234]:
#es la funcion donde pone el contienente a cada pais
def get_continent(country_name):
    try:
        country_code = pc.country_name_to_country_alpha2(country_name, cn_name_format="default")
        continent_code = pc.country_alpha2_to_continent_code(country_code)
        continent_map = {
            "AF": "Africa",
            "AS": "Asia",
            "EU": "Europe",
            "NA": "North America",
            "SA": "South America",
            "OC": "Oceania",
            "AN": "Antarctica"
        }
        return continent_map.get(continent_code, "Unknown")
    except:
        return "Unknown"
    
#creamos una funcion que añada la nueva columna indicando el contienente
def add_continent_column(df_Calidad_vida):
    df_Calidad_vida["Continent"] = df_Calidad_vida["country"].apply(get_continent)
    return df_Calidad_vida

In [235]:
#Agregamos la columna de contienente a nuestro data frame
df_Calidad_vida = add_continent_column(df_Calidad_vida)

In [236]:
df_Calidad_vida.head()

country  Purchasing Power Value Purchasing Power Category  \
0    Afghanistan                   32.15                'Very Low'   
1  Aland Islands                  125.01               'Very High'   
2        Albania                   42.82                     'Low'   
3       Alderney                    0.00                   Unknown   
4        Algeria                   27.60                'Very Low'   

   Safety Value Safety Category  Health Care Value Health Care Category  \
0         25.33           'Low'              24.24                'Low'   
1         71.81          'High'              79.72               'High'   
2         55.52      'Moderate'              48.21           'Moderate'   
3         83.79     'Very High'             100.00          'Very High'   
4         47.54      'Moderate'              54.43           'Moderate'   

   Climate Value Climate Category  Cost of Living Value  \
0           0.00          Unknown                 21.08   
1           0.00          Unknown                 53.44   
2          86.43      'Very High'                 40.85   
3           0.00          Unknown                  0.00   
4          94.82      'Very High'                 25.31   

  Cost of Living Category Property Price to Income Value  \
0              'Very Low'                            7.8   
1                   'Low'                           5.33   
2                   'Low'                          14.88   
3                 Unknown                            0.0   
4              'Very Low'                           21.7   

  Property Price to Income Category  Traffic Commute Time Value  \
0                             'Low'                       56.17   
1                             'Low'                       19.05   
2                            'High'                       36.74   
3                       'Very High'                        5.00   
4                       'Very High'                       45.09   

  Traffic Commute Time Category  Pollution Value Pollution Category  \
0                   'Very High'            84.44        'Very High'   
1                    'Very Low'            18.05         'Very Low'   
2                    'Moderate'            77.25             'High'   
3                    'Very Low'             1.72         'Very Low'   
4                        'High'            63.87             'High'   

  Quality of Life Value Quality of Life Category Continent  
0                   0.0                  Unknown      Asia  
1                   0.0                  Unknown   Unknown  
2            ': 104.16'                    'Low'    Europe  
3                   0.0                  Unknown   Unknown  
4             ': 98.83'               'Very Low'    Africa

In [237]:
df_Calidad_vida["Continent"].value_counts()

Continent
Africa           53
Asia             48
Europe           46
North America    30
Unknown          24
Oceania          21
South America    14
Name: count, dtype: int64

In [238]:
df_nan_continente = df_Calidad_vida[df_Calidad_vida["Continent"] == "Unknown"]
lista_paises_sin_continente = df_nan_continente["country"].tolist()

In [239]:
lista_paises_sin_continente

['Aland Islands',
 'Alderney',
 'Antigua And Barbuda',
 'Bosnia And Herzegovina',
 'Curacao',
 'French Southern Territories',
 'Hong Kong (China)',
 'Isle Of Man',
 'Kosovo (Disputed Territory)',
 'Macao (China)',
 'Reunion',
 'Saint Helena',
 'Saint Kitts And Nevis',
 'Saint Vincent And The Grenadines',
 'Saint-Pierre And Miquelon',
 'Sao Tome And Principe',
 'Sint Maarten',
 'Timor-Leste',
 'Trinidad And Tobago',
 'Turks And Caicos Islands',
 'Us Virgin Islands',
 'Vatican City',
 'Wallis And Futuna',
 'Western Sahara']

Continentes:
Aland Islands – Europa
Alderney – Europa
Antigua and Barbuda – América (Caribe)
Bosnia and Herzegovina – Europa
Curacao – América (Caribe)
French Southern Territories – Antártida (Territorio dependiente de Francia)
Hong Kong (China) – Asia
Isle of Man – Europa
Kosovo (Disputed Territory) – Europa (Controversia política)
Macao (China) – Asia
Reunion – África (Territorio francés en el Océano Índico)
Saint Helena – África (Territorio británico en el Atlántico sur)
Saint Kitts and Nevis – América (Caribe)
Saint Vincent and the Grenadines – América (Caribe)
Saint-Pierre and Miquelon – América (Territorio francés en el Atlántico)
Sao Tome and Principe – África
Sint Maarten – América (Caribe)
Timor-Leste – Asia
Trinidad and Tobago – América (Caribe)
Turks and Caicos Islands – América (Caribe)
US Virgin Islands – América (Caribe, Territorio estadounidense)
Vatican City – Europa
Wallis and Futuna – Oceanía (Territorio francés en el Pacífico)
Western Sahara – África (Disputa territorial con Marruecos, reconocido por algunos países como territorio independiente)

In [240]:
# Diccionario con países y sus continentes en inglés

continentes_dict = {
    'Aland Islands': 'Europe',
    'Alderney': 'Europe',
    'Antigua And Barbuda': 'North America',  # Caribe, parte de América del Norte
    'Bosnia And Herzegovina': 'Europe',
    'Curacao': 'North America',  # Caribe, parte de América del Norte
    'French Southern Territories': 'Antarctica',
    'Hong Kong (China)': 'Asia',
    'Isle Of Man': 'Europe',
    'Kosovo (Disputed Territory)': 'Europe',
    'Macao (China)': 'Asia',
    'Reunion': 'Africa',
    'Saint Helena': 'Africa',
    'Saint Kitts And Nevis': 'North America',  # Caribe, parte de América del Norte
    'Saint Vincent And The Grenadines': 'North America',  # Caribe, parte de América del Norte
    'Saint-Pierre And Miquelon': 'North America',  # Caribe, parte de América del Norte
    'Sao Tome And Principe': 'Africa',
    'Sint Maarten': 'North America',  # Caribe, parte de América del Norte
    'Timor-Leste': 'Asia',
    'Trinidad And Tobago': 'South America',  # Caribe, pero más cerca de América del Sur
    'Turks And Caicos Islands': 'North America',  # Caribe, parte de América del Norte
    'Us Virgin Islands': 'North America',  # Caribe, parte de América del Norte
    'Vatican City': 'Europe',
    'Wallis And Futuna': 'Oceania',
    'Western Sahara': 'Africa',
}

# Actualizar la columna 'Continent' en el DataFrame
df_Calidad_vida['Continent'] = df_Calidad_vida['country'].map(continentes_dict).fillna(df_Calidad_vida['Continent'])


In [241]:
df_Calidad_vida["Continent"].value_counts()

Continent
Africa           57
Europe           52
Asia             51
North America    38
Oceania          22
South America    15
Antarctica        1
Name: count, dtype: int64

In [242]:
df_nan_continente = df_Calidad_vida[df_Calidad_vida["Continent"] == "Unknown"]
lista_paises_sin_continente = df_nan_continente["country"].tolist()
lista_paises_sin_continente 


[]

In [243]:
df_Calidad_vida.to_csv(".//files/Quality_of_Life_limpio.csv", index="False")